packages/@webex/plugin-meetings/src/breakouts/index.ts

```diff
@@ -0,0 +1,206 @@
+   * get a breakout.roster event when you join the meeting
+   */
+  queryRosters() {
+    this.webex.request(
+      {uri: `${this.url}/roster`, qs: {locusUrl: btoa(this.locusUrl)}
+    }).then((result) => {
+      const {body: {rosters}} = result;
+      
+      rosters.forEach(({locus}) => {
+        this.handleRosterUpdate(locus);
+      })
+
+      this.trigger(BREAKOUTS.EVENTS.MEMBERS_UPDATE);
+    })
+  },
+
+  handleRosterUpdate(locus) {
+    const sessionId = locus.controls?.breakout?.sessionId;
+
+    const session = this.breakouts.get(sessionId);
+
+    if (!session) {
+      return;
+    }
+
+    session.parseRoster(locus);
+  },
+
+  /**
+   * Sets up listener for broadcast messages sent to the breakout session
+   */
+  listenToBroadcastMessages() {
+    this.listenTo(this.webex.internal.llm, 'event:breakout.message', (event) => {
+      const {data: {senderUserId, sentTime, message}} = event;
+
+      this.trigger(BREAKOUTS.EVENTS.MESSAGE, {
+        senderUserId,
+        sentTime,
+        message,
+        // FIXME: This is only the current sessionId
+        // We'd need to check that the dataChannelUrl is still the same
+        // to guarantee that this message was sent to this session
+        sessionId: this.currentBreakoutSession.sessionId,
+      });
+    })
+  },
+
+  /**
+   * Sets up a listener for roster messags from mecury
+   */
+  listenToBreakoutRosters() {
+    this.listenTo(this.webex.internal.mercury, 'event:breakout.roster', (event) => {
+      this.handleRosterUpdate(event.data.locus);
+      this.trigger(BREAKOUTS.EVENTS.MEMBERS_UPDATE);
+    })
+  },
+
+  /**
+   * Updates the information about the current breakout
+   * @param {Object} params
+   * @returns {void}
+   */
+  updateBreakout(params) {
+    this.set(params);
+
+    this.set('currentBreakoutSession', {
+      sessionId: params.sessionId,
+      groupId: params.groupId,
+      name: params.name,
+      current: true,
+      sessionType: params.sessionType,
+      url: params.url,
+      active: false,
+      allowed: false,
+      assigned: false,
+      assignedCurrent: false,
+      requested: false,
+    });
+  },
+
+  /**
+   * Updates the information about available breakouts
+   * @param {Object} payload
+   * @returns {void}
+   */
+  updateBreakoutSessions(payload) {
+    const breakouts = {};
+
+    if (payload.breakoutSessions) {
+      forEach(['active', 'assigned', 'allowed', 'assignedCurrent', 'requested'], (state) => {
+        forEach(payload.breakoutSessions[state], (breakout) => {
+          const {sessionId} = breakout;
+
+          if (!breakouts[sessionId]) {
+            breakouts[sessionId] = breakout;
+            breakouts[sessionId].active = false;
+            breakouts[sessionId].assigned = false;
+            breakouts[sessionId].allowed = false;
+            breakouts[sessionId].assignedCurrent = false;
+            breakouts[sessionId].requested = false;
+          }
+
+          breakouts[sessionId][state] = true;
```

Comment:
oh, I see what you're doing. 
You're updating it here...

Seems a bit confusing?

you could just do something like 
```js
 breakouts[sessionId].allowed = state === 'allowed';
```
 in the if statement above.
 
 Also, would be nice to create constants for all the available states.

Reply:
The payload arrives as an object with arrays of sessions in different states
e.g.
{
active: [{session1}, {session2}],
allowed: [{session1}]
}

I want to flatten and deduplicate this into a single collection of models where each state is just a flag. Sessions can appear in multiple places, so you can't just assign say allowed based on whether the current bucket is allowed because you may be overwriting a value set in by a previous bucket.

Reply:
I'm now using constants for session states

Reply:
makes sense!!